README.md
## 章节介绍：

* **1. 概述：** 介绍背景。
* **2. Demo 使用：** 介绍 Demo 如何跑起来。包括环境搭建，编译下载，设备控制。
* **3. 开发指南：** 介绍 Demo 如何二次开发。包括文件结构，功能介绍，API介绍，资源占用。
* **4. 相关链接：** 给出与 joylink 相关的链接。包括 Demo 下载，微联文档。

## 1. 概述
ESP 平台实现了京东微联 Joylink2.1.22 协议。用户可以参考 Espressif 提供的设备端 Demo 进行二次开发，快速接入京东微联云平台。

Demo 参考京东官方 [Joylink2.1.22 SDK](https://smartdev.jd.com/docCenterDownload/list/2)，添加了 WiFi 相关、OTA、Flash 存储等适配层，用户只需要关心少部分 API，如初始化、数据收发、事件回调等，加快了用户的二次开发速度。适配层具有一定的通用性且开源，用户可根据产品需求自行修改。
Demo 使用的通讯方式为非透传模式。非透传模式下需要在开发者中心上传空实现的 lua 脚本： `only_trans.lua`。

目前京东微联已经停止支持 SmartConfig 的配网方式，所以 Demo 中删除 SmartConfig 配网，目前支持 SoftAp 和 Thunder 配网，芯片上电默认进入 SoftAP 配网，如果用户需要使用 Thunder 配网，需要购买京东的音箱，音箱需要进入沙箱模式，与音箱相关的设置，用户可以咨询京东。

## 2. Demo 使用
用户拿到乐鑫提供的 esp-joylink 后，根据不同的 example 编译下载固件到乐鑫 ESP8266 、 ESP32 或者 ESP32-S2 开发板。使用京东微联 APP 扫描测试设备的二维码进行配置。配置激活成功后便可进行设备控制。此 Demo 对应的测试设备类型为“智能家居 / 生活电器 / 灯具”。

### 2.1. 环境搭建

* 环境准备
  * **ESP32**：[开发环境准备](https://docs.espressif.com/projects/esp-idf/en/stable/get-started/)
  * **ESP32-S2**：[开发环境准备](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/get-started/index.html)
  * **ESP8266**：[开发环境准备](https://github.com/espressif/ESP8266_RTOS_SDK/blob/release/v3.0/docs/en/get-started/get-started-devkitc.rst)
  
* 硬件准备  
  * **开发板**：ESP8266 、ESP32-S2 与 ESP32 开发板（[ESP-DevKitC](http://espressif.com/zh-hans/company/contact/buy-a-sample)）
  * **路由器**：使用 2.4GHz 网络，可以连接外网
  * **手机**：安装[京东微联沙箱 APP](https://smartdev.jd.com/docCenterDownload/list/2)

* 软件准备
  
* **Demo下载**：下载 Demo 工程 [esp-joylink](https://github.com/espressif/esp-joylink.git)
  
* Demo 的网络拓扑结构

    <img src="docs/_picture/环境搭建.jpg" width = "500" alt="i2c hardware" align=center />

### 2.2 编译下载

* 工程编译
    首先设置  `IDF_PATH`  的路径

    * ESP8266 平台，需使用 ESP8266_RTOS_SDK v3.2 及之后的版本，推荐使用 ESP8266_RTOS_SDK v3.3 
    * ESP32 平台， 需使用 ESP-IDF v4.0 及之后的版本，推荐使用 ESP-IDF v4.0 和 v4.2
    
    * ESP32-S2 平台，需使用 ESP-IDF v4.2 及之后的版本，推荐使用 ESP-IDF v4.2 
    
    然后在 menuconfig 里面配置 joylink 相关参数 `UUID`，`CID`，`PID`， `PUBLIC_KEY`，`PRIVATE_KEY`，`MAC_ADDRESS`，其中 `UUID`、`CID` 和 `PID` 分别代表产品的唯一标识码、品类和品牌，在云端产品中的位置如下
    
    <img src="docs/_picture/UUID、CID和PID.jpg" width = "900" alt="i2c hardware" align=center />
    
    `PUBLIC_KEY ` 为产品公钥，显示在设备端开发页面，设备认证采用一机一密的方式 ，用户需导入 `MAC_ADDRESS` (设备 MAC 地址) 来生成 `PRIVATE_KEY` (设备私钥) ，具体生成方式请在设备端开发页面点击 “导入 MAC 地址生成私钥” 选项并按提示进行。
    
    <img src="docs/_picture/公钥、私钥和MAC.jpg" width = "900" alt="i2c hardware" align=center />
    
    以light_demo为例，我们提供了默认的sdkconfig.defaults及partiton table文件（light_demo/default_configs目录下），用户可根据芯片不同，进行使用和参考。
    
    进入example/light_demo目录，然后输入命令 `$IDF_PATH/tools/idf.py build`， 编译工程。对于 ESP32-S2 平台，执行 `build`  命令之前应先执行`$IDF_PATH/tools/idf.py set-target esp32s2 ` 命令，切换为 ESP32-S2 的编译环境，之后再 `$IDF_PATH/tools/idf.py build`  编译工程。
    
    (如果使用 ESP8266_RTOS_SDK v3.2，请使用 `make`  命令进行编译工程)

* 固件烧写  
    1. 如果没有安装串口驱动，需要先安装[串口驱动](http://www.usb-drivers.org/ft232r-usb-uart-driver.html)。
    2. 输入命令 `$IDF_PATH/tools/idf.py -p PORT flash`  (ESP8266_RTOS_SDK v3.2 使用 `make flash ESPPORT=PORT` ) ，将编译好的 bin 文件烧录到开发板中，`PORT`  选择和配置请参考[文档说明](https://docs.espressif.com/projects/esp-idf/en/stable/get-started/index.html#step-6-connect-your-device)。

对于esp32平台，Demo中包含两种配网方式: SoftAP 配网和 BLE 配网，可在 menuconfig 的 joylink 组件配置中进行选择切换，默认是使用SoftAP 配网

<img src="docs/_picture/使能BLE.jpg" width = "800" alt="i2c hardware" align=center />

若切换 BLE 配网，首先需使能  `Whether or not use Joylink ble configure wifi` 选项，然后要在云端产品的基本信息页面中选择 "京东BLE配网" 并保存，最后重新编译和烧写后方可使用。

<img src="docs/_picture/配网选择.jpg" width = "850" alt="i2c hardware" align=center />

## 3. 开发指南
本章详细介绍 Demo 工程，如需二次开发请关注本章。
用户需要调用的 API 和参数配置相关的头文件在 `port/include/esp_joylink.h` 中。

### 3.1 文件结构

    .
    ├── CMakeLists.txt
    ├── component.mk
    ├── docs
    │   ├── md                                            //  md files
    │   │   └── 量产说明.md
    │   └── _picture
    │
    ├── config
    │   └── mass_mfg
    │       ├── multipule_mfg_config.csv
    │       ├── multipule_mfg_values.csv
    │       └── single_mfg_config.csv
    │   
    ├── examples
    │   ├── light_demo                                    //  light_demo
    │   │   ├── CMakeLists.txt
    │   │   ├── components
    │   │   ├── default_configs                           // default sdkconfigs
    │   │   ├── main
    │   │   ├── README.md
    │   │   └── Makefile
    │   └── transparent_transmission
    │   
    ├── joylink_dev_sdk                                   // joylink SDK source code
    │   ├── ble_sdk
    │   ├── example
    │   ├── joylink
    │   ├── lua
    │   ├── Makefile
    │   ├── pal
    │   ├── README
    │   └── scripts
    │       
    ├── Kconfig
    ├── port                                              // joylink SDK adaptation
    │   ├── include
    │   ├── joylink_ble.c
    │   └── joylink_upgrade.c
    │
    └── README.md


### 3.2 参数配置

* 产品信息配置  
    在 `memuconfig` 中， 用户可以修改以下这些参数，如 AES_KEY, UUID, PID等，系统会调用 `joylink_dev_init()` 传入产品注册的信息，注册事件回调函数。用户如果需要修改为自己的设备，首先需要在京东开发者中心上申请自己的设备，并根据相应设备信息在此处修改。  
    因为这些参数需要在服务器后台产品基本信息里获取，或者向京东有关部门咨询。

### 3.3 配网
京东配网方式目前主推 SoftAP(自主激活), Demo 中同时支持 SoftAP(自主激活+代理激活)、Thunder 配网, 默认设置是进入 SoftAP 自主激活, 如果需要代理激活，可以在 Makefile 中将自主激活的宏关闭。

<img src="docs/_picture/配网方式后台配置界面.jpg" width = "500" alt="i2c hardware" align=center />

### 3.4 固件升级
实现固件升级需要上传新的固件到服务器后台。并填写正确的版本信息。
**注意**： 厂商必须严格按照固定格式填写版本号，如（1.1.1.20151020_release）。
如下图，后台设置在线升级界面。序号是上传 bin 文件后自动生成的。设备端上传的 version 如果与这个版本号不同， APP 会自动推送 OTA 的提示。

<img src="docs/_picture/ota_后台版本设置.png" width = "900" alt="i2c hardware" align=center />

### 3.5量产功能

每台设备对应唯一的UUID、DeviceMAC、PrivateKey，用户除了可以在`menuconfig`配置相关参数信息，也可以通过烧录NVS分区的方式进行量产。

具体请参考[量产说明](docs/md/量产说明.md)

>如果执行了`idf.py erase_flash`, 需要重新烧录三元组

## 4. 相关链接

* Espressif 官网： http://espressif.com
* ESP32 SDK 下载： https://github.com/espressif/esp-idf
* ESP8266 SDK 下载： https://github.com/espressif/ESP8266_RTOS_SDK
* 烧录工具：http://espressif.com/en/support/download/other-tools
* 京东微联官网：https://smartcloud.jd.com
* 官方JoyLink协议SDK： https://storage.jd.com/testsmartcloud/joylink_dev_sdk.zip
* 设备入网的协议文档： https://storage.jd.com/testsmartcloud/JoylinkConfigDoc.zip


joylink_dev_sdk/RELEASENOTES

--------------------------------------------------------------------------------
Version     : 2.1.22.21
Date        : 2020年 10月 15日 星期四 11:46:53 CST
Owner       : liyanming10
Description : 
1、实现连接退避机制。
2、露出子设备状态上报接口。

--------------------------------------------------------------------------------
Version     : 2.1.22.20
Date        : 2020年 08月 11日 星期二 10:30:58 CST
Owner       : liyanming10
Description : 
1、收数据去除1400字节的限制。
2、解决IOT通道收数据不完整的bug。

--------------------------------------------------------------------------------
Version     : 2.1.22.19
Date        : 2020年 07月 23日 星期四 11:33:43 CST
Owner       : liyanming10
Description : 
1、完成自动化编译sdk脚本。
2、解决云菜谱返回的biz_code不正确的问题。
3、alive埋点日志增加离线次数参数上报。
4、mac的buffer长度统一改成17，防止有些设备是16字节时会崩溃。
5、jl_platform_send的实现send参数使用从time_out改为flags
6、解决局域网控制的一个bug，复用了时间戳校验函数的返回值，会导致局域网控制子设备失败。

--------------------------------------------------------------------------------
Version     : 2.1.22
Date        : 2020年 04月 14日 星期二 19:55:29 CST
Owner       : liyanming10
Description : 
1、代码中标准库函数的调用改为pal层调用。
2、解决认证超时导致设备重新建立长连接再次认证。
3、加版本上报功能。
4、日志埋点上报线程栈空间从0x4000改为0x1000.
5、露出激活绑定信息接口；露出局域网激活绑定开关接口（默认关闭激活绑定，防止被别人绑走）。
6、代码优化，删掉user_jlp。
7、x86环境下，把OTA升级的源码编译到jt中。
8、优化ota升级逻辑，crc校验边下载数据边校验。
9、SoftAP单次配网，保证pubKey和R2不变，防止重发时生成的sessionKey不一致。
10、WiFi应答包打印日志不匹配（udp 打印成tcp了）。
11、解决SoftAP配网上报日志点位不正确的问题。
12、支持通过__JOYLINK_LOG_SUPPORT_HTTP__宏切换激活绑定日志上报的协议，如果定义，则通过http上报，没有定义通过https上报。
13、加通过__JOYLINK_ALL_ACTIVE_INFO__宏定义控制激活绑定接口。
14、解决切换到activeAndBindForDevice接口激活绑定失败的问题。

--------------------------------------------------------------------------------
Version     : 2.1.21
Date        : 2020年03月13日 10:17:39
Owner       : liyanming10
Description : 
1、增加设备上电埋点。
2、删除_AGENT_GW_宏控的代码。
3、connect时上报失败次数，失败之后延时3秒开始再次建立连接。
4、细化接收数据返回值埋点，区分失败，服务端断开，大数据包长度超限。
5、优化控制结果上报。
6、增加定时（10分钟一次）上报alive日志，确保主任务活着。
7、增加热点断开埋点，热点断开之后计时，连接之后把断开的时间上报。
8、解决静态扫描的bug，解决ssid为空的bug，会影响一键配网。
9、softap激活绑定之后将token和url返回给app，以验证其正确性。
10、其它代码优化，增加可读性。
11、解决线程不同步导致的日志埋点中参数ip不能实时更新的问题。
12. 解决一个日志模块计数bug带来的内存泄漏的问题。
13. 解决断网解析域名内存泄漏的问题。
14. 埋点上报线程增加网络判断和配网模式判断，配网模式不上报日志，断网时1秒上报一次，每上报一次日志休眠50ms。
15. 优化日志上报逻辑，当上报失败时，适当的增加休眠时间。
16. 修改joylink_main_loop末尾的sleep，解决局域网高频压测不通过的问题。
17. softAP配网函数增加超时机制，超时时间由传入的参数决定。
18. OTA升级url buffer的长度改为200字节，防止文件名过长导致内存越界。
19. 解决庆科平台有配网失败的问题。主要是增加client_fd >= 0的判断。
20. 解决兼容64bits的问题。
22. 解决代码检测工具扫描出的bug。

--------------------------------------------------------------------------------
Version     : 2.0.21
Date        : Fri Jan 10 17:37:39 2020 +0800
Owner       : liyanming10
Description : 
1、解决AUTH时长连接断开，不会重新建立连接，除非断电重启的问题。
2、解决套接字overflow的bug，体现在长连接成功了，但是设备无法接受云端下发的数据。
3、内存优化，优化堆空间，节省堆空间约18KB。
4、增加日志埋点的功能，方便排查设备离线问题。

--------------------------------------------------------------------------------
VERSION:2.0.7
HEAD:35fe9dd7501bf5754b8e7e5951ac527926dcd159
DATE:04/24/18

--------------------------------------------------------------------------------

Upgrade urgency levels:

LOW:      No need to upgrade unless there are new features you want to use.
MODERATE: Program an upgrade of the server, but it's not urgent.
HIGH:     There is a critical bug that may affect a subset of users. Upgrade!
CRITICAL: There is a critical bug affecting MOST USERS. Upgrade ASAP.

--------------------------------------------------------------------------------
--[ joylink 1.3.3 ] Release date: 

# UPGRADE URGENCY: LOW for joylink. 

* [NEW] No. 
* [FIX] Add CID firmwareVersion modelCode in func joylink_package_scan 
* [FIX] Add productuuid in func joylink_parse_write_key 
=======================

** IMPORTANT ** 


--[ joylink 1.3.3 ] Release date: 

# UPGRADE URGENCY: LOW for joylink. 

* [NEW] No. 
* [FIX] joylink_json.c:joylink_parse_scan, add  cJSON_Delete(pJson);
* [FIX] joylink_json.c:joylink_package_scan, del cJson_CreateArry;
=======================

** IMPORTANT ** 

--------------------------------------------------------------------------------
--[ joylink 1.3.2 ] Release date: 

# UPGRADE URGENCY: LOW for joylink. 

* [NEW] No. 
* [FIX] Add sdk/lua/example.lua sdk/lua/only_trans.lua 
* [FIX] Change var name in joylink.h     
=======================

** IMPORTANT ** 

--------------------------------------------------------------------------------
--[ joylink 1.3.1 ] Release date: 

# UPGRADE URGENCY: LOW for joylink. 

* [NEW] Add big package cut and send ,but code is for debug. 

* [FIX] 
=======================

** IMPORTANT ** 

--------------------------------------------------------------------------------

--[ joylink 1.3.0 ] Release date: 

# UPGRADE URGENCY: CRITICAL for joylink. 

* [FIX] Fix the error of joylink_parse_lan_write_key()

* [FIX] Change .jlp.trantype = 0, to .jlp.trantype = 1

Migrating from 1.2 to 1.3
=========================

joylink_dev_sdk 1.2 is mostly a strict subset of 1.3. However there are a few things
that you should be aware of:

The following default attr changed:

    * .jlp.trantype = 1, means suport lua script.
--------------------------------------------------------------------------------

joylink_dev_sdk 1.2 release notes
=======================

** IMPORTANT ** Check the 'Migrating from 1.1 to 1.2' section at the end of
                this file for information about what changed between 1.1 and
                1.2 and how this may affect your application.

--------------------------------------------------------------------------------
Upgrade urgency levels:

LOW:      No need to upgrade unless there are new features you want to use.
MODERATE: Program an upgrade of the server, but it's not urgent.
HIGH:     There is a critical bug that may affect a subset of users. Upgrade!
CRITICAL: There is a critical bug affecting MOST USERS. Upgrade ASAP.
--------------------------------------------------------------------------------

--[ joylink 1.2.0 ] Release date: 14 Jul 2014

# UPGRADE URGENCY: LOW for joylink , this is a features enhancement
                   release mostly. 

* [FIX] Recive data that's len bigger than 1400 from server. (steven)


user/app_joylink.c
/*
 * ESPRESSIF MIT License
 *
 * Copyright (c) 2018 <ESPRESSIF SYSTEMS (SHANGHAI) PTE LTD>
 *
 * Permission is hereby granted for use on ESPRESSIF SYSTEMS ESP8266 only, in which case,
 * it is free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished
 * to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

#include "esp_common.h"
#include "freertos/semphr.h"
#include "esp_joylink.h"
#include "joylink_aes.h"
#include "esp_joylink_log.h"
#include "button/button.h"
#include "status_led/led.h"
#include "uart.h"

static const char *TAG = "esp8266_joylink";

#define JOYLINK_WIFI_CFG_BUTTON_NUM       (13) /*!< set button pin, restart to wifi config */
#define JOYLINK_FACTORY_RESET_BUTTON_NUM  (14) /*!< set button pin, clear joylink config info */
#define LED_IO_NUM                        (15) /*!< set led gpio, led status */

#define JSON_VALUE_STR_LEN  (32) /*!< json value string length */

/*
 * light demo parameters:
 * power       |  int     |  0/1
 * brightness  |  int     |  0 ~ 100
 * colortemp   |  int     |  3000 ~ 6500
 * mode        |  int     |  0/1/2/3/4
 * color       |  string  |  string of color number
*/

typedef struct {
    char power;
    char light_value;
    int  temp_value;
    char work_mode;
    char color[10];
} virtual_device_t;

static led_handle_t led_0;
static virtual_device_t virtual_device;
static xSemaphoreHandle xSemWriteInfo = NULL;
static xSemaphoreHandle xSemReadInfo = NULL;

static char *device_attr[5] = { "power", "brightness", "colortemp", "mode", "color"};

const char *joylink_json_upload_str = "{\"code\":%d,\"streams\":[\
{\"current_value\": \"%d\",\"stream_id\": \"power\"},\
{\"current_value\": \"%d\",\"stream_id\": \"brightness\"},\
{\"current_value\": \"%d\",\"stream_id\": \"colortemp\"},\
{\"current_value\": \"%d\",\"stream_id\": \"mode\"},\
{\"current_value\": \"%s\",\"stream_id\": \"color\"}]}";

static uint8_t is_server;

void user_show_rst_info(void)
{
    struct rst_info *rtc_info = system_get_rst_info();
    JOYLINK_LOGI("reset reason: %x", rtc_info->reason);

    if ((rtc_info->reason == REASON_WDT_RST)
            || (rtc_info->reason == REASON_EXCEPTION_RST)
            || (rtc_info->reason == REASON_SOFT_WDT_RST)) {

        if (rtc_info->reason == REASON_EXCEPTION_RST) {
            JOYLINK_LOGI("Fatal exception (%d):", rtc_info->exccause);
        }

        JOYLINK_LOGI("dbg: epc1=0x%08x, epc2=0x%08x, epc3=0x%08x, excvaddr=0x%08x, depc=0x%08x",
                   rtc_info->epc1, rtc_info->epc2, rtc_info->epc3, rtc_info->excvaddr, rtc_info->depc);
    }
}

void get_free_heap_task()
{
    while (1) {
        JOYLINK_LOGD("free RAM size: %d", system_get_free_heap_size());
        vTaskDelay(2000 / portTICK_RATE_MS);
    }
}

int cloud_read(virtual_device_t* virtual_dev, int ms_wait)
{
    JOYLINK_PARAM_CHECK(!virtual_dev);

    int i, valueLen = 0;
    char data_buff[JSON_VALUE_STR_LEN];
    char *down_cmd = (char *)malloc(JOYLINK_DATA_LEN);

    if (esp_joylink_read(down_cmd, JOYLINK_DATA_LEN, ms_wait) < 0) {
        JOYLINK_LOGE("esp_joylink_read is error");
        return E_RET_ERROR;
    }

    for (i = 0; i < 5; i++) {
        char *valueStr = joylink_json_get_value_by_id(down_cmd, strlen(down_cmd), device_attr[i], &valueLen);

        if (valueStr && valueLen > 0) {
            memset(data_buff, 0, JSON_VALUE_STR_LEN);
            memcpy(data_buff, valueStr, valueLen);
            JOYLINK_LOGD("valueStr:%s, attrLen:%d", data_buff, valueLen);

            switch (i) {
                case 0:
                    virtual_dev->power = atoi(data_buff);
                    break;

                case 1:
                    virtual_dev->light_value = atoi(data_buff);
                    break;

                case 2:
                    virtual_dev->temp_value = atoi(data_buff);
                    break;

                case 3:
                    virtual_dev->work_mode = atoi(data_buff);
                    break;

                case 4:
                    strcpy(virtual_dev->color, data_buff);
                    break;

                default:
                    break;
            }
        }
    }

    if (NULL != down_cmd) {
        free(down_cmd);
    }

    return E_RET_OK;
}

int cloud_write(virtual_device_t virtual_device, int ms_wait)
{
    char *up_cmd = (char *)malloc(JOYLINK_DATA_LEN);
    memset(up_cmd, 0, JOYLINK_DATA_LEN);
    sprintf(up_cmd, joylink_json_upload_str, 0, virtual_device.power, virtual_device.light_value,
            virtual_device.temp_value, virtual_device.work_mode, virtual_device.color);
    int ret = esp_joylink_write(up_cmd, strlen(up_cmd)+1, ms_wait);
    free(up_cmd);
    return ret;
}

void read_task_test(void *arg)
{
    int cnt = 0;

    for (;;) {
        xSemaphoreTake(xSemReadInfo, portMAX_DELAY);
        cloud_read(&virtual_device, portMAX_DELAY);

        if (cnt++ % 2) {
            led_state_write(led_0, LED_NORMAL_ON);
        } else {
            led_state_write(led_0, LED_NORMAL_OFF);
        }

        JOYLINK_LOGD("highwater %d", uxTaskGetStackHighWaterMark(NULL));
    }
}

void write_task_test(void *arg)
{
    for (;;) {
        xSemaphoreTake(xSemWriteInfo, portMAX_DELAY);
        int ret = cloud_write(virtual_device, 500 / portTICK_RATE_MS);

        if (ret == E_RET_ERROR) {
            JOYLINK_LOGE("esp_write is err");
        }

        vTaskDelay(500 / portTICK_RATE_MS);
        JOYLINK_LOGD("highwater %d", uxTaskGetStackHighWaterMark(NULL));
    }
}

int joylink_event_handler(joylink_action_t action)
{
    joylink_event_param_t *param = NULL;

    if (action.arg) {
        param = ((joylink_event_param_t*)action.arg);
    }

    switch (action.event) {
        case JOYLINK_EVENT_WIFI_START_SMARTCONFIG:
            JOYLINK_LOGD("JOYLINK_EVENT_WIFI_START_SMARTCONFIG");
            led_state_write(led_0, LED_SLOW_BLINK);
            break;

        case JOYLINK_EVENT_WIFI_START_SOFTAP_CONFIG:
            JOYLINK_LOGD("JOYLINK_EVENT_WIFI_START_SOFTAP_CONFIG");
            led_state_write(led_0, LED_SLOW_BLINK);
            break;

        case JOYLINK_EVENT_WIFI_GOT_IP:
            JOYLINK_LOGD("JOYLINK_EVENT_WIFI_GOT_IP");
            led_state_write(led_0, LED_QUICK_BLINK);
            JOYLINK_LOGD("conn reason:%d, ssid:%s, password:%s", param->conn.rsn,
                    param->conn.conf.ssid, param->conn.conf.password);
            break;

        case JOYLINK_EVENT_CLOUD_CONNECTED:
            JOYLINK_LOGD("JOYLINK_EVENT_CLOUD_CONNECTED");
            led_state_write(led_0, LED_NORMAL_ON);
            break;

        case JOYLINK_EVENT_CLOUD_DISCONNECTED:
            log_warn("JOYLINK_EVENT_CLOUD_DISCONNECTED");
            break;

        case JOYLINK_EVENT_GET_DEVICE_DATA:
            JOYLINK_LOGD("JOYLINK_EVENT_GET_DEVICE_DATA");
            JOYLINK_LOGD("must to post cmd, max len:%d, %s, %s", \
                      param->get.max_len, param->get.is_json ? "json" : "script", \
                      param->get.is_server_cmd ? "server" : "lan");
            is_server = param->get.is_server_cmd;
            xSemaphoreGive(xSemWriteInfo);
            break;

        case JOYLINK_EVENT_SET_DEVICE_DATA:
            JOYLINK_LOGD("JOYLINK_EVENT_SET_DEVICE_DATA");
            JOYLINK_LOGD("must to read cmd, cmd len:%d, %s, %s", \
                      param->set.cmd_len, param->set.is_json ? "json" : "script", \
                      param->set.is_server_cmd ? "server" : "lan");
            is_server = param->get.is_server_cmd;
            xSemaphoreGive(xSemReadInfo);
            break;

        case JOYLINK_EVENT_OTA_START:
            JOYLINK_LOGD("JOYLINK_EVENT_OTA_START");
            JOYLINK_LOGD("start to ota, ver: %d, url: %s", param->ota.version, param->ota.url);
            break;

        default:
            break;
    }

    return E_RET_OK;
}

static void initialise_key(void)
{
    button_handle_t btn_handle1 = button_dev_init(JOYLINK_WIFI_CFG_BUTTON_NUM, 1, BUTTON_ACTIVE_LOW);
    button_dev_add_tap_cb(BUTTON_TAP_CB, joylink_reset_to_smartconfig, "PUSH", 50 / portTICK_RATE_MS, btn_handle1);
    button_dev_add_press_cb(0, joylink_reset_to_softap_innet, "long press", 3000 / portTICK_RATE_MS, btn_handle1);

    button_handle_t btn_handle2 = button_dev_init(JOYLINK_FACTORY_RESET_BUTTON_NUM, 0, BUTTON_ACTIVE_LOW);
    button_dev_add_tap_cb(BUTTON_PUSH_CB, joylink_restore_factory, "PUSH", 50 / portTICK_RATE_MS, btn_handle2);
}

static void initialise_led(void)
{
    led_0 = led_create(LED_IO_NUM, LED_DARK_LOW);
    led_state_write(led_0, LED_NORMAL_OFF);
}
#define SNTP_OPMODE_POLL            0

static void initialize_sntp(void)
{
    sntp_setoperatingmode(SNTP_OPMODE_POLL);
    sntp_setservername(0, "pool.ntp.org");
    sntp_init();
}

int joylink_start(void)
{
    joylink_info_t *product_info = (joylink_info_t*)malloc(sizeof(joylink_info_t));
    
    if (!product_info) {
        JOYLINK_LOGE("malloc fail\n");
    }
    
    memset(product_info,0x0,sizeof(joylink_info_t));
    
    product_info->innet_aes_key = JOYLINK_AES_KEY;
    product_info->device_info.local_port = JOYLINK_LOCAL_PORT;
    product_info->device_info.jlp.version = JOYLINK_VERSION;
    product_info->device_info.jlp.devtype = JOYLINK_DEVTYPE;
    product_info->device_info.jlp.cmd_tran_type = E_CMD_TYPE_LUA_SCRIPT,
    strncpy(product_info->device_info.jlp.joylink_server, JOYLINK_SERVER,sizeof(product_info->device_info.jlp.joylink_server));
    product_info->device_info.jlp.server_port = JOYLINK_SERVER_PORT;
    strncpy(product_info->device_info.jlp.firmwareVersion, JOYLINK_FW_VERSION, sizeof(product_info->device_info.jlp.firmwareVersion));
    strncpy(product_info->device_info.jlp.modelCode, JOYLINK_MODEL_CODE, sizeof(product_info->device_info.jlp.modelCode));
    strncpy(product_info->device_info.jlp.uuid, JOYLINK_UUID, sizeof(product_info->device_info.jlp.uuid));
    product_info->device_info.jlp.lancon = E_LAN_CTRL_ENABLE;
    strncpy(product_info->device_info.idt.cloud_pub_key, CLOUD_PUB_KEY, sizeof(product_info->device_info.idt.cloud_pub_key));

    strncpy(product_info->device_info.jlp.CID, JOYLINK_CID, sizeof(product_info->device_info.jlp.CID));
    
    uint8 macaddr[6];
    char mac_str[18] = {0};
    wifi_get_macaddr(0, macaddr);
    snprintf(product_info->device_info.jlp.mac, sizeof(product_info->device_info.jlp.mac), MACSTR, MAC2STR(macaddr));

    JOYLINK_LOGI("*********************************");
    JOYLINK_LOGI("*         JOYLINK INFO          *");
    JOYLINK_LOGI("*********************************");
    JOYLINK_LOGI("UUID     : %s", product_info->device_info.jlp.uuid);
    JOYLINK_LOGI("mac      : %s", product_info->device_info.jlp.mac);
    JOYLINK_LOGI("type     : %d", product_info->device_info.jlp.devtype);
    JOYLINK_LOGI("version  : %d", product_info->device_info.jlp.version);
    JOYLINK_LOGI("fw_ver   : %s", product_info->device_info.jlp.firmwareVersion);
    JOYLINK_LOGI("model    : %s", product_info->device_info.jlp.modelCode);
    JOYLINK_LOGI("tran type: %d", product_info->device_info.jlp.cmd_tran_type);
    JOYLINK_LOGI("innet key: %s", product_info->innet_aes_key);
    JOYLINK_LOGI("public key: %s", product_info->device_info.idt.cloud_pub_key);
    JOYLINK_LOGI("sys run  : %s", system_upgrade_userbin_check() ? "user2" : "user1");
    JOYLINK_LOGI("*********************************");

    esp_joylink_event_init(joylink_event_handler);
    initialise_key();
    initialise_led();
    esp_info_init();
    initialize_sntp();
    esp_joylink_init(product_info);
    esp_joylink_set_version(JOYLINK_VERSION);

    if (xSemWriteInfo == NULL) {
        xSemWriteInfo = (xSemaphoreHandle)xSemaphoreCreateBinary();
    }

    if (xSemReadInfo == NULL) {
        xSemReadInfo = (xSemaphoreHandle)xSemaphoreCreateBinary();
    }

    xTaskCreate(get_free_heap_task, "get_free_heap_task", 128, NULL, 5, NULL);
    xTaskCreate(read_task_test, "read_task_test", 256, NULL, 5, NULL);
    xTaskCreate(write_task_test, "write_task_test", 256, NULL, 4, NULL);

    if (!product_info) {
        free(product_info);
    }
    
    return E_RET_OK;
}

extern void joylink_init_aes_table(void);

void user_demo(void)
{
    JOYLINK_LOGI("*********************************************");
    JOYLINK_LOGI("* user.bin compiled @ %s %s *", __DATE__, __TIME__);
    JOYLINK_LOGI("* free RAM size %d ", system_get_free_heap_size());
    user_show_rst_info();
    JOYLINK_LOGI("*********************************************");

    joylink_init_aes_table();
    joylink_start();  /* start joylink */
}
